### PR TITLE
Implement most Bmi1/Bmi2 intrinsics

### PR DESCRIPTION
**Pull request by tannergooding**
This is 7/10 APIs for https://github.com/dotnet/coreclr/issues/18712

### CODE REVIEW

**Comment by tannergooding** [comment]
FYI. @fiigii, @CarolEidt, @eerhardt

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by tannergooding** [comment]
Updated the new scalar test template to run the "NotSupported" scenario for `long`/`ulong` on 32-bit machines.

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
CI appears to be back up. Going to try to get all the jobs queued again.

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
Pretty sure this requires another change (making BMI1/BMI2 dependent on COMPlus_EnableAVX, due to their using the VEX encoding)...

File: src/jit/emitxarch.cpp [hunk 1]
```diff
@@ -50,6 +50,38 @@ bool IsFMAInstruction(instruction ins)
...
+bool IsBMIInstruction(instruction ins)
+{
+    return (ins >= INS_FIRST_BMI_INSTRUCTION) && (ins <= INS_LAST_BMI_INSTRUCTION);
+}
+
+regNumber getBmiRegNumber(instruction ins)
+{
+    switch (ins)
+    {
+        case INS_blsi:
```
Comment (tannergooding) [inline src/jit/emitxarch.cpp]: These (and bzhi, which isn't included yet) have special encoding

File: src/jit/emitxarch.cpp [hunk 2]
```diff
@@ -810,7 +854,33 @@ unsigned emitter::emitOutputRexOrVexPrefixIfNeeded(instruction ins, BYTE* dst, c
...
                 {
                     case 0x66:
-                        vexPrefix |= 0x01;
+                        if (IsBMIInstruction(ins))
+                        {
+                            switch (ins)
+                            {
+                                case INS_pdep:
+                                {
+                                    vexPrefix |= 0x03;
```
Comment (tannergooding) [inline src/jit/emitxarch.cpp]: The BMI instructions are SSE38 encoded (4-byte encoding) but have a different prefix.

File: tests/src/JIT/HardwareIntrinsics/X86/Shared/GenerateTests.csx [hunk 3]
```diff
@@ -786,6 +786,100 @@ private static readonly (string templateFileName, Dictionary<string, string> tem
...
+    ("ScalarUnOpTest.template",  new Dictionary<string, string> { ["Isa"] = "Bmi1", ["Method"] = "TrailingZeroCount",       ["RetBaseType"] = "UInt32", ["Op1BaseType"] = "UInt32",                             ["NextValueOp1"] = "(uint)(random.Next(0, int.MaxValue))",                                                              ["ValidateResult"] = "uint expectedResult = 0; for (int index = 0; ((data >> index) & 1) == 0; index++) { expectedResult++; } isUnexpectedResult = (expectedResult != result);" }),
+    ("ScalarUnOpTest.template",  new Dictionary<string, string> { ["Isa"] = "Bmi1", ["Method"] = "TrailingZeroCount",       ["RetBaseType"] = "UInt64", ["Op1BaseType"] = "UInt64",                             ["NextValueOp1"] = "(ulong)(random.Next(0, int.MaxValue))",                                                             ["ValidateResult"] = "ulong expectedResult = 0; for (int index = 0; ((data >> index) & 1) == 0; index++) { expectedResult++; } isUnexpectedResult = (expectedResult != result);" }),
+};
+
+private const string ValidateBmi2ParallelBitDepositUInt32 = @"uint temp = left;
```
Comment (tannergooding) [inline tests/src/JIT/HardwareIntrinsics/X86/Shared/GenerateTests.csx]: Couldn't think of a way to validate the functionality without implementing the software version...

Comment (tannergooding) [reply tests/src/JIT/HardwareIntrinsics/X86/Shared/GenerateTests.csx]: Should probably attribute the Intel Architecture manual here as well, as this is following the "Operation" pseudo-code listed.

Comment (CarolEidt) [reply tests/src/JIT/HardwareIntrinsics/X86/Shared/GenerateTests.csx]: Unfortunate to have to do this, but I'm not sure of a better way ...

Comment (tannergooding) [reply tests/src/JIT/HardwareIntrinsics/X86/Shared/GenerateTests.csx]: Added a comment indicating this is based on the operation pseudo-code from the architecture manual.

**Comment by tannergooding** [comment]
Updated the BMI1/2 instruction sets to be dependent on the COMPlus_EnableAVX flag

**Comment by tannergooding** [comment]
ci.dot.net is still down, so this will need for tests to run.

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
Applied formatting patch.

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x64 Checked jitnox86hwintrinsic please

@dotnet-bot test Windows_NT x86 Checked jitincompletehwintrinsic please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Windows_NT x86 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Windows_NT x86 Checked jitnox86hwintrinsic please

@dotnet-bot test Ubuntu x64 Checked jitincompletehwintrinsic please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnoavx2 please
@dotnet-bot test Ubuntu x64 Checked jitx86hwintrinsicnosimd please
@dotnet-bot test Ubuntu x64 Checked jitnox86hwintrinsic please

**Comment by tannergooding** [comment]
Looks like the `CoreFX Tests` are unrelated and failing in other PRs as well...
> MESSAGE:
Assert.Throws() Failure\r\nExpected: typeof(System.ArgumentNullException)\r\nActual: typeof(System.AccessViolationException): Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
+++++++++++++++++++
STACK TRACE:
at System.Runtime.InteropServices.Marshal.ReadValueSlow[T](Object ptr, Int32 ofs, Func`3 readValueHelper) at System.Drawing.Font.ToLogFont(Object logFont, Graphics graphics) at System.Drawing.Font.ToLogFont(Object logFont)

@CarolEidt, do you know if this is a known issue? Am I good to merge?

**Comment by CarolEidt** [comment]
It seems that the same corefx test is failing in all the current PRs. I wouldn't worry about that one, but would probably wait for the other runs to finish, and make sure that any failures are the same as on other PRs. I'll check to see if there's an open issue on the ` System.Drawing.Common.Tests` failure.

**Comment by tannergooding** [comment]
> but would probably wait for the other runs to finish

Will do. The `x64 Release CoreFX Tests` have actually already finished and just didn't update. The Ubuntu and OSX10.12 ones actually kicked off when I commented, so I'll wait until those complete (and validate any failures) before merging.

**Comment by CarolEidt** [comment]
I filed #18840 for the corefx failure.

**Comment by AaronRobinsonMSFT** [comment]
I thought that @stephentoub addressed this kind of issue in [corefx#30830](https://github.com/dotnet/corefx/issues/30830), but it might be failing here also.

**Pull request merged**